src/jockey/geometry.rs

```diff
@@ -0,0 +1,252 @@
+}
+
+/// A struct represents a geometry.
+pub struct Geometry {
+    /// Count of vertices.
+    pub count: GLsizei,
+
+    /// Drawing mode e.g. `gl::TRIANGLES` .
+    pub mode: GLenum,
+
+    /// Attributes of the geometry. Keys are attribute location.
+    pub attributes: HashMap<GLuint, GeometryAttribute<GLfloat>>,
+
+    /// Index buffer of the geometry.
+    pub indices: Option<GeometryAttribute<GLuint>>,
+
+    /// A vao object for this geometry.
+    _vao: Option<GLuint>,
+}
+
+impl<T> GeometryAttribute<T> where T: std::fmt::Debug {
+    pub fn init(
+        array: Vec<T>,
+        size: GLuint,
+        type_: GLenum,
+    ) -> Self {
+        GeometryAttribute {
+            array,
+            size,
+            type_,
+            normalized: gl::FALSE,
+            target: gl::ARRAY_BUFFER,
+            usage: gl::STATIC_DRAW,
+            _buffer: None,
+        }
+    }
+
+    /// Make a vertex buffer object out of this attribute and assign it to its buffer field.
+    pub fn buffer(&mut self) -> GLuint {
+        match self._buffer {
+            None => {
+                let mut buffer = 0;
+
+                unsafe {
+                    gl::GenBuffers(1, &mut buffer);
+                    gl::BindBuffer(self.target, buffer);
+                    gl_debug_check!();
+                }
+
+                unsafe {
+                    gl::BufferData(
+                        self.target,
+                        (self.array.len() * std::mem::size_of::<T>()) as _,
+                        std::mem::transmute(self.array.as_ptr()),
+                        self.usage,
+                    );
+                    gl_debug_check!();
+                }
+
+                self._buffer = Some(buffer);
+
+                buffer
+            },
+            Some(buffer) => buffer,
+        }
+    }
+
+    pub fn vertex_attrib_pointer(&mut self, index: GLuint) {
+        match self._buffer {
+            None => (),
+            Some(buffer) => {
+                unsafe {
+                    gl::BindBuffer(self.target, buffer);
+                    gl_debug_check!();
+                }
+
+                unsafe {
+                    gl::VertexAttribPointer(
+                        index,
+                        self.size as _,
+                        self.type_,
+                        self.normalized,
+                        0,
+                        std::ptr::null(),
+                    );
+                    gl_debug_check!();
+                }
+            }
+        }
+    }
+
+    /// Delete the buffer object.
+    pub fn delete_buffer(&mut self) {
+        match self._buffer {
+            None => (),
+            Some(buffer) => {
+                unsafe {
+                    gl::DeleteBuffers(1, &buffer);
+                    // gl_debug_check!();
+                }
+
+                self._buffer = None;
+            }
+        }
+    }
+}
+
+impl<T> std::fmt::Debug for GeometryAttribute<T> where T: std::fmt::Debug {
```

Comment:
This `Debug` implementation can be generated by the compiler by adding `#[derive(Debug)]` in front of the struct definition

Reply:
I see! this is good